### PR TITLE
Fix race condition in ReleaseOnDetach test

### LINKED ISSUE #47098: Test failed: profiler/unittest/releaseondetach/releaseondetach.sh
Job : [runtime-coreclr gcstress0x3-gcstress0xc 20210117.1](https://dev.azure.com/dnceng/public/_build/results?buildId=954737&view=ms.vss-test-web.build-test-results-tab)

Failed tests:
```
CoreCLR Linux x64 Checked gcstress0x3 @ Ubuntu.1804.Amd64.Open
        - profiler/unittest/releaseondetach/releaseondetach.sh
```

Error message:
```
Unhandled exception. System.Exception: Profilee returned exit code 139 instead of expected exit code 100.
   at Profiler.Tests.ProfilerTestRunner.FailFastWithMessage(String error)
   at Profiler.Tests.ProfilerTestRunner.Run(String profileePath, String testName, Guid profilerClsid, String profileeArguments, ProfileeOptions profileeOptions)
   at Profiler.Tests.ReleaseOnShutdown.Main(String[] args)
/home/helixbot/work/A20108CB/w/A00F0887/e/profiler/unittest/releaseondetach/releaseondetach.sh: line 395: 20532 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"

Return code:      1
Raw output file:      /home/helixbot/work/A20108CB/w/A00F0887/e/profiler/unittest/Reports/profiler.unittest/releaseondetach/releaseondetach.output.txt
Raw output:
BEGIN EXECUTION
/home/helixbot/work/A20108CB/p/corerun releaseondetach.dll ''
Profiler path: /home/helixbot/work/A20108CB/w/A00F0887/e/profiler/unittest/releaseondetach/libProfiler.so
Attaching profiler /home/helixbot/work/A20108CB/w/A00F0887/e/profiler/unittest/releaseondetach/libProfiler.so to self.
Profiler.dll!DllGetClassObject
Profiler.dll!Profiler::Initialize
Initialize for attach started
RequestProfilerDetach successful
Profiler detach succeeded
PROFILER TEST PASSES
Gathering state for process 20543 corerun
Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.20543.dmp
Written 76877824 bytes (18769 pages) to core file
Dump successfully written

Test failed: Profilee returned exit code 139 instead of expected exit code 100.
Gathering state for process 20532 corerun
Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.20532.dmp
Written 97529856 bytes (23811 pages) to core file
Dump successfully written
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/home/helixbot/work/A20108CB/p
> /home/helixbot/work/A20108CB/w/A00F0887/e/profiler/unittest/releaseondetach/releaseondetach.sh

stack trace:
   at profiler_unittest._releaseondetach_releaseondetach_._releaseondetach_releaseondetach_sh()
```

### PR DESCRIPTION
**Pull request by davmason**
Fixes #47098

There was a race condition in the ReleaseOnDetachTest. The call from the managed testcase to request a profiler attach to itself will block until the profiler succeeds in attaching, then the profiler requests a detach instantly. The vast majority of the time it takes a bit for the detach to actually happen, but sometimes the detach could happen before the call to SetBoolPtr happens. Then SetBoolPtr would call in to the profiler after detach and potentially AV.

The fix here is to add synchronization by polling for _doneFlag.

### CODE REVIEW

**Review by hoyosjs — APPROVED** [review]

File: src/tests/profiler/native/releaseondetach/releaseondetach.cpp [hunk 1]
```diff
@@ -51,6 +53,17 @@ ReleaseOnDetach::~ReleaseOnDetach()
...
 
+
+    for (int i = 0; i < 50000; ++i)
+    {
+        if  (_doneFlag != NULL)
+        {
+            break;
+        }
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```
Comment (sdmaclea) [inline src/tests/profiler/native/releaseondetach/releaseondetach.cpp:64]: These types of timeouts tend to be fragile.  Is this long enough for gcstress runs?

**Review by sdmaclea — APPROVED** [review]

File: src/tests/profiler/native/releaseondetach/releaseondetach.cpp (hunk 1 above)
Comment (davmason) [reply src/tests/profiler/native/releaseondetach/releaseondetach.cpp]: It has been running with a timeout for a while and this is the first failure of this type I've seen so I was trying to be lazy. I should just fix it the right way though, I just pushed a commit getting rid of all the sleeps.

File: src/tests/profiler/unittest/releaseondetach.cs [hunk 2]
```diff
@@ -58,19 +39,12 @@ public unsafe static int RunTest(string[] args)
...
-            // The native part just sets it.
-            #pragma warning disable CS0420
-            fixed (bool *boolPtr = &_profilerDone)
-            {
-                PassBoolToProfiler(boolPtr);
-
-                WaitForDetach();
-            }
+            PassCallbackToProfiler(() => _profilerDone.Set());
+            _profilerDone.WaitOne();
```
Comment (hoyosjs) [inline src/tests/profiler/unittest/releaseondetach.cs:47]: I like this a lot better. But we should probably have a timeout here.

Comment (davmason) [reply src/tests/profiler/unittest/releaseondetach.cs:47]: I'm open to adding the timeout, just want to know your reasoning. Are you concerned about the test getting in to an unknown state and taking forever for the testcase to time out?

Comment (hoyosjs) [reply src/tests/profiler/unittest/releaseondetach.cs:47]: Yeah. The "this hung for some reason" case, just so it doesn't fireball all the tests that run along with this one. A ridiculous timeout is fine, say 5 min.

**Pull request merged**